contributors/design-proposals/node/optional-service-links.md

```diff
@@ -0,0 +1,52 @@
+# Optional service links
+
+Status: Pending
+
+Version: Alpha | Beta | GA
+
+Implementation Owner: @kongslund
+
+## Motivation
+
+Today, a list of all services that were running when a pod's containers are created is automatically injected to those containers as environment variables matching the syntax of Docker links. There is no way to disable this.
+
+Docker links have long been considered as a [deprecated legacy feature](https://docs.docker.com/engine/userguide/networking/default_network/dockerlinks/) of Docker since the introduction of networks and DNS. Likewise, in Kubernetes, DNS is to be preferred over service links.
+
+Possible issues with injected service links are
+
+* Accidental coupling.
+* Incompatibilities with container images that no longer utilize service links and explicitly fail at startup time if certain service links are defined.
+* Performance penalty in starting up pods [for namespaces with many services](https://github.com/kubernetes/kubernetes/issues/1768#issuecomment-330778184).
+
+## Proposal
+
+Make it possible for a user to disable injection of service links into containers of a pod by adding a disable flag to the pod's spec. Make the default value false in order to stay backwards compatible with the v1 API. Make an exception for the `kubernetes` service in the master namespace so that it will always get injected.
+
+## User Experience
+
+### Use Cases
+
+* As a user, I want to be able to disable service link injection since the injected environment variables interfere with a Docker image that I am trying to run on Kubernetes
+* As a user, I want to be able to disable service link injection since I don't need it and it takes increasingly longer time to start pods as services are added to the namespace.
+
+## Implementation
+
+`PodSpec` is extended with an additional field, `enableServiceLinks`. The field should be a pointer to a boolean and default to true if nil.
```

Comment:
This may need to be disableServiceLinks, so that the existing behavior would be in effect if it were the zero value.
cc @erictune

Reply:
Either is fine with me. However, there seems to be a preference for `enable` instead of `disable` cf. https://github.com/kubernetes/community/pull/1249#discussion_r147019407

Reply:
Brian, as a pointer zero value would be interpreted as existing behavior by older clients - can you articulate a bit more what you’d be concerned about as a bool ptr?

Reply:
Even if we go with `enable`, we can set its default value to `true`, so I don't see any backwards compatibility issues there.

Reply:
Sorry for delay. I do want this to happen.

If we make it a pointer and omitempty, will both true and false values be explicitly encoded as long as the field is non-null? I want to ensure that podspecs without the field can be unambiguously interpreted.

Also, doing this will add a field to every Pod and every pod template rendered. It will change podspec hashes, but that should be ok. It will make all pods more verbose, but maybe more users will become aware of the field and use it.

If we could, it would be great to not apply the default in pod templates. In addition to not changing hashes, this would make it feasible to set the field via admission control.

Reply:
Now that we use versioned clients leaving fields unset on the wire has fewer downsides - a client already has to test existence, so not applying the default on the apiserver is reasonable (effectively how runAsUser behaves).  We definitely want to define the default, but we don’t have to fill it except at point of use.  The biggest risk is clients inconsistently interpreting the default, which our current “always default” strategy minimizes.  But as the apply discussion shows, it’s not necessarily an advantage for end users.

Maybe explicit defaulting is something a client should be able to request from the server, although we could only really change that behavior for existing fields until a v2.

Reply:
I don't think that's the case, to what in the apply discussion are you referring?

Client-applied defaults are ~impossible to ever change in our system, even over version boundaries.

Reply:
I still think that defaults explicitly manifested by the server is the right thing. It just needs to be done where the resources are actually reified, not in templates.

Reply:
So, do we have some agreement there? Are we OK with the field named `disableServiceLinks`, pointer, omitempty, with the default `false` value on the server side? That seems to address @bgrant0607's doubts - it will preserve the old behavior by default and would not require to define that field.

Reply:
`enable` is probably a better API.  I think that's what Brian was saying - pointers make this possible.

Any serialized form of an object which has this value specified will retain that value.  Anything that has this value omitted will get `nil` which will be defaulted to whatever the versioned default is.

Right?

So if you serialize it in v1, when you read it back it will deserialize and default.  If you write it back as v2, that v1 default is still retained...

Am I missing something?

Reply:
OK, I agree. Let's go with `enable`. I will change that in my implementation.